pytorch_lightning/trainer/data_loading.py

```diff
@@ -114,33 +119,34 @@ def auto_add_worker_init_fn(self, dataloader: DataLoader) -> None:
             dataloader.worker_init_fn = partial(pl_worker_init_function, rank=self.global_rank)
 
     def auto_add_sampler(self, dataloader: Any, shuffle: bool, mode: Optional[RunningStage] = None) -> Any:
-        # don't do anything if it's not a dataloader
-        is_dataloader = isinstance(dataloader, DataLoader)
-        # don't manipulate iterable datasets
-        is_iterable_ds = has_iterable_dataset(dataloader)
-
         if isinstance(dataloader, CombinedLoader):
+            # apply `auto_add_sampler` on all the collection of loaders
             dataloader.loaders = apply_to_collection(dataloader.loaders, DataLoader, self.auto_add_sampler, shuffle)
             return dataloader
 
-        if not is_dataloader or is_iterable_ds:
+        # don't do anything if it's not a dataloader
+        if not isinstance(dataloader, DataLoader):
             return dataloader
 
-        need_dist_sampler = self.accelerator_connector.is_distributed and not isinstance(
-            dataloader.sampler, DistributedSampler
-        )
-        if self.accelerator_connector.replace_sampler_ddp and need_dist_sampler:
+        if (
+            self.accelerator_connector.replace_sampler_ddp
+            and self.accelerator_connector.is_distributed
+            and not isinstance(dataloader.sampler, DistributedSampler)
+            and not has_iterable_dataset(dataloader)
+        ):
             if not isinstance(dataloader.sampler, (SequentialSampler, RandomSampler)):
                 raise MisconfigurationException(
                     "You seem to have configured a sampler in your DataLoader. This will be replaced "
                     " by `DistributedSampler` since `replace_sampler_ddp` is True and you are using"
                     " distributed training. Either remove the sampler from your DataLoader or set"
                     " `replace_sampler_ddp`=False if you want to use your custom sampler."
                 )
-
-            # replace with distributed sampler
             sampler = self._get_distributed_sampler(dataloader, shuffle, mode=mode)
-            dataloader = self.replace_sampler(dataloader, sampler, mode=mode)
+        else:
+            # use current sampler
+            sampler = dataloader.sampler
+
+        dataloader = self.replace_sampler(dataloader, sampler, mode=mode)
```

Comment:
Hi @tchaton, should we have replaced the sampler even when `self.accelerator_connector.replace_sampler_ddp` is `False`? I'm not familiar with the logic here, but a few of our tests failed because `replace_sampler` is now invoked after this change. Specifically, the test uses a custom sampler, but `_resolve_batch_sampler` expects to be able to re-construct the sampler following way:
```python
if (batch_sampler is not None and type(batch_sampler) is not BatchSampler) or is_predicting:
    batch_sampler = type(batch_sampler)(
        sampler,
        batch_size=batch_sampler.batch_size,
        drop_last=(False if is_predicting else batch_sampler.drop_last),
    )
```
However, the custom sampler in question doesn't have the expected fields.

Reply:
It won't actually replace it if `replace_sampler_ddp`. But the code should be refactored as it does more than just replace the sampler.